proc/threads.go

```diff
@@ -145,73 +146,167 @@ func topframe(g *G, thread *Thread) (Stackframe, error) {
+	if err != nil && stepInto {
+		return err
+	}
 
-// Set breakpoints at every line, and the return address. Also look for
-// a deferred function and set a breakpoint there too.
-func (dbp *Process) next(g *G, topframe Stackframe) error {
-	pcs := dbp.lineInfo.AllPCsBetween(topframe.FDE.Begin(), topframe.FDE.End()-1, topframe.Current.File)
+	cond := sameGoroutineCondition(dbp.SelectedGoroutine)
 
-	var deferpc uint64 = 0
-	if g != nil && g.DeferPC != 0 {
-		_, _, deferfn := dbp.goSymTable.PCToLine(g.DeferPC)
-		var err error
-		deferpc, err = dbp.FirstPCAfterPrologue(deferfn, false)
-		if err != nil {
-			return err
+	if stepInto {
+		for _, instr := range text {
+			if instr.Loc.File != topframe.Current.File || instr.Loc.Line != topframe.Current.Line || !instr.IsCall() {
+				continue
+			}
+
+			if instr.DestLoc != nil && instr.DestLoc.Fn != nil {
+				if err := dbp.setStepIntoBreakpoint([]AsmInstruction{instr}, cond); err != nil {
+					dbp.ClearTempBreakpoints()
+					return err
+				}
+			} else {
+				// Non-absolute call instruction, set a StepBreakpoint here
+				if _, err := dbp.SetTempBreakpoint(instr.Loc.PC, StepBreakpoint, cond); err != nil {
+					if _, ok := err.(BreakpointExistsError); !ok {
+						dbp.ClearTempBreakpoints()
+						return err
+					}
+				}
+			}
 		}
 	}
 
-	var covered bool
-	for i := range pcs {
-		if topframe.FDE.Cover(pcs[i]) {
-			covered = true
-			break
+	if !csource {
+		deferreturns := []uint64{}
+
+		// Find all runtime.deferreturn locations in the function
+		// See documentation of Breakpoint.DeferCond for why this is necessary
+		for _, instr := range text {
+			if instr.IsCall() && instr.DestLoc != nil && instr.DestLoc.Fn != nil && instr.DestLoc.Fn.Name == "runtime.deferreturn" {
+				deferreturns = append(deferreturns, instr.Loc.PC)
+			}
 		}
-	}
 
-	if !covered {
-		fn := dbp.goSymTable.PCToFunc(topframe.Ret)
-		if g != nil && fn != nil && fn.Name == "runtime.goexit" {
-			return nil
+		// Set breakpoint on the most recently deferred function (if any)
+		var deferpc uint64 = 0
+		if dbp.SelectedGoroutine != nil && dbp.SelectedGoroutine.DeferPC != 0 {
+			_, _, deferfn := dbp.goSymTable.PCToLine(dbp.SelectedGoroutine.DeferPC)
+			var err error
+			deferpc, err = dbp.FirstPCAfterPrologue(deferfn, false)
+			if err != nil {
+				return err
+			}
+		}
+		if deferpc != 0 && deferpc != topframe.Current.PC {
+			bp, err := dbp.SetTempBreakpoint(deferpc, NextBreakpoint, cond)
+			if err != nil {
+				if _, ok := err.(BreakpointExistsError); !ok {
+					dbp.ClearTempBreakpoints()
+					return err
+				}
+			}
+			if bp != nil {
+				bp.DeferCond = true
+				bp.DeferReturns = deferreturns
+			}
 		}
 	}
-	if deferpc != 0 {
-		pcs = append(pcs, deferpc)
+
+	// Add breakpoints on all the lines in the current function
+	pcs := dbp.lineInfo.AllPCsBetween(topframe.FDE.Begin(), topframe.FDE.End()-1, topframe.Current.File)
+
+	if !csource {
+		var covered bool
```

Comment:
I know this is carry over, but it would be nice to have some comments here explaining this logic.